fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAutoCompleteResultCursor.java

```diff
@@ -169,12 +170,325 @@ private void performLookup() throws IOException {
+        try {
+            IndexableField primaryKey = searcher.doc(scoreDoc.doc).getField(LuceneIndexMaintainer.PRIMARY_KEY_FIELD_NAME);
+            BytesRef pk = primaryKey.binaryValue();
+            return state.store.loadRecordAsync(Tuple.fromBytes(pk.bytes)).thenApply(rec -> {
+                if (rec == null) {
+                    // No document found. Return original record.
+                    return null;
+                }
+                // Extract the indexed fields from the document again
+                final List<LuceneDocumentFromRecord.DocumentField> documentFields = LuceneDocumentFromRecord.getRecordFields(state.index.getRootExpression(), rec)
+                        .get(groupingKey == null ? TupleHelpers.EMPTY : groupingKey);
+
+                // Search each field to find the first match.
+                final int maxTextLength = Objects.requireNonNull(state.context.getPropertyStorage()
+                        .getPropertyValue(LuceneRecordContextProperties.LUCENE_AUTO_COMPLETE_TEXT_SIZE_UPPER_LIMIT));
+                @Nullable LuceneDocumentFromRecord.DocumentField matchedField = null;
+                @Nullable String matchedText = null;
+                for (LuceneDocumentFromRecord.DocumentField documentField : documentFields) {
+                    Object fieldValue = documentField.getValue();
+                    if (fieldValue instanceof String) {
+                        String text = (String) fieldValue;
+                        if (text.length() > maxTextLength) {
+                            // Apply the text length filter before searching through the text for the
+                            // matched terms
+                            continue;
+                        }
+                        String match = searchAllMaybeHighlight(text, queryTokens, prefixToken, highlight);
+                        if (match != null) {
+                            matchedField = documentField;
+                            matchedText = match;
+                            break;
+                        }
+                    }
+                }
+
+                if (matchedField == null) {
+                    return null;
+                }
+                Tuple key = Tuple.from(matchedField.getFieldName(), matchedText);
+                if (groupingKey != null) {
+                    key = groupingKey.addAll(key);
+                }
+                // TODO: Add the primary key to the index entry
+                // Not having the primary key is fine for auto-complete queries that just want the
+                // text, but queries wanting to do something with both the auto-completed text and the
+                // original record need to do something else
+                IndexEntry indexEntry = new IndexEntry(state.index, key, Tuple.from(scoreDoc.score));
```

Comment:
is it expected to use the doc's score rather than the concept "weights", and get rid of the logic to re-calculate the weights based on the positions of the matches in their texts and sort the results based on that?